File: Data Structures & Algorithms/balanced-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int recurse(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int i = recurse(root.left);
        int j = recurse(root.right);
        return 1 + Math.max(i,j);
    }
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        int i = recurse(root.right);
        int j = recurse(root.left);
        if (Math.abs(i - j) > 1) return false;

        return isBalanced(root.left) && isBalanced(root.right);
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        result = 0
        stack = [(root, root.val)]

        while stack:
            node, biggest_so_far = stack.pop()
            if node.val >= biggest_so_far:
                result += 1
                biggest_so_far = node.val
            if node.right: 
                stack.append((node.right, biggest_so_far))
            if node.left:
                stack.append((node.left, biggest_so_far))
        return result

        